gold/Gold_data.hpp

```diff
@@ -0,0 +1,271 @@
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+#include <vector>
+#include <utility>
+#include <iostream>
+#include <algorithm>
+#include <functional>
+
+#include "lrand.hpp"
+
+class Gold_data {
```

Comment:
<bikeshedding>

AFAICT, `Gold_data` is basically just a sparse byte array with a way to efficiently iterate through the elements that are present.  It might be beneficial to explain it in these terms.

The implementation tries to be more efficient by organizing the sparse array as a collection of dense intervals with and growing the existing intervals if we add new data right before or right after.  This recalls the K&R malloc implementation and is a pitfall-rich landscape.

It seems to have an awful lot of API for something that only tries to do a few things: read(interval), write(interval, data), merge(gold_data), and iterate(function).  It seems like much of the API is internal but it's not clear which.